core/src/main/java/com/microsoft/applicationinsights/internal/channel/common/ApacheSender43.java

```diff
@@ -91,27 +68,44 @@ public void close() {
         try {
             ((CloseableHttpClient)getHttpClient()).close();
         } catch (IOException e) {
-            InternalLogger.INSTANCE.error("Failed to close http client, exception: %s", e.toString());
+            InternalLogger.INSTANCE.error("Failed to close http httpClient, exception: %s", e.toString());
+        } finally {
+            synchronized (lock) {
+                if (connectionManager != null) {
+                    connectionManager.shutdown();
+                }
+            }
         }
     }
 
     @Override
     public HttpClient getHttpClient() {
-        if (!isClientInitialized) {
-            synchronized (this) {
-                if (!isClientInitialized) {
-                    while (httpClientRef.get() == null) {
-                        try {
-                            TimeUnit.MILLISECONDS.sleep(3);
-                        } catch (InterruptedException e){
-                        }
-                    }
-                    isClientInitialized = true;
+        CloseableHttpClient result = httpClient;
+        if (result == null) {
+            synchronized (lock) {
+                connectionManager = initializeConnectionManager();
+                result = httpClient;
+                if (result == null) {
+                    httpClient = result = initializeClient(connectionManager);
```

Comment:
This seems very confusing. Please simplify these assignments.

Reply:
Assignments evaluate from right to left. It's equivalent to `httpClient = (result = initializeClient(connectionManager))`

Reply:
Yes I understand that. Taking a look at the practice you mentioned it looks reasonable. I would recommend adding the link to the stackoverflow link in the doc so future developers get aware.

Reply:
@littleaj yes I have no problems myself reading this. I am looking from the eyes of future inherriter of the code base  and it looks invasive :)